WIP towards make PyCall use Conda if system python is not found

dhoegh opened the pull request:
This is a WIP towards figuring out how Conda can be used as a replacement of system python and it is parallel work to https://github.com/JuliaLang/IJulia.jl/pull/357.
I still haven't figured out the API for how PyPlot/PyCall/IJulia should interact with Conda and if they should require Conda or if the user should add Conda.

stevengj commented:
The difficulty with this is that many people using PyCall will be Python users who want to call Python packages that they've installed by their familiar way in Python.   If PyCall uses Conda's Python, they will be surprised because it won't find any of the packages they've installed with their regular Python.

dhoegh commented:
Yes that is the reasoning about not requiring Conda, but if Conda is installed then it prefers the python distribution in Conda. This means the changes will not affect people having a system wide python and do not have Conda installed.

dhoegh commented:
I have updated the PR. The miniconda python distribution in Conda only takes up 123 MB when nothing is installed which acceptable. I have added a `use_conda` constant for the purpose of automatically installation of python dependencies when Conda is used. This is shown in https://github.com/stevengj/PyPlot.jl/pull/170.

stevengj commented on test/runtests.jl:
```diff
@@ -1,4 +1,7 @@
 using Base.Test, PyCall, Compat
+import Conda
+
+PyCall.use_conda && Conda.add("numpy")
```
Probably `Conda.add("numpy")` should be done in `Pkg.build`.

stevengj commented on deps/build.jl:
```diff
@@ -121,7 +122,17 @@ end
-const python = get(ENV, "PYTHON", "python")
+use_conda = false
+const python = try 
+    pyconfigvar(get(ENV, "PYTHON", "python"),"VERSION","")
```
Probaby call `get(ENV, "PYTHON", "python")` only once, and check for Python >= 2.7.   i.e.
```jl
let py = get(ENV, "PYTHON", "python"), vers = VersionNumber(pyconfigvar(py,"VERSION","0.0"))
    if vers < v"2.7"
        error("Python version $vers < 2.7 is not supported")
    end
end
```

stevengj commented on deps/build.jl:
```diff
@@ -121,7 +122,17 @@ end
+    info( "No system-wide Python was found; got the following error:\n",
+    "$e1\nusing the Python distribution in the Conda package")
+    use_conda = true
+    abspath(Conda.PREFIX, "python" * (@windows? ".exe" : ""))
```
Shouldn't that be `Conda.BINDIR`?

dhoegh replied:
No, Conda.BINDIR points to the binaries of installed packages and not the python binaries.

dhoegh replied:
I am tired of windows/linux differences. On linux it should be `abspath(Conda.BINDIR, "python"))` on windows it should be `abspath(Conda.PREFIX, "python".exe"))`. This almost dictates another constant in Conda called `PYTHONDIR` else we should adjust PyCall in this special case.

stevengj commented:
Would be good to have Travis testing too.   Maybe just add Python 2.6 to the Travis tests, if that is allowed by Travis.  With the version check I suggested above, that will force Pkg.build to use Conda.

dhoegh commented:
I have addressed your comments. Maybe you could cancel the old appveyor build to reduce the load on appveyor.

stevengj commented:
cancelled the build.

stevengj commented:
Travis is failing with `could not spawn `/home/travis/.julia/v0.3/Conda/deps/usr/python -c "import distutils.sysconfig; print(distutils.sysconfig.get_config_var('VERSION'))"`: no such file or directory (ENOENT)`

Probably, `python` is in `Conda.BINDIR` on non-Windows systems.

(I don't understand why they couldn't just use the same paths on Unix and Windows.)

dhoegh commented:
I have made a PR to `Conda.jl` to add `PYTHONDIR`. I will comeback tomorrow and to finish.

Luthaf commented:
> (I don't understand why they couldn't just use the same paths on Unix and Windows.)

Because they reproduce the way Python is packaged on Windows. But I have no idea why it was done so for Python in the first place.

tkelman commented:
endless frustration caused by someone deciding to be different on different platforms decades ago

dhoegh commented:
Everything is passing now.

stevengj commented:
Thanks!
